Add webauthn blog post (#1690)
FusionAuth/fusionauth-site: site/_posts/2022-11-01-authenticators-ceremonies-webauthn-oh-my.md

* It's easier to use since users have to manage fewer passwords.
* It eliminates the need for servers to store passwords. With passwords, if a server happens to be compromised, the user might be at risk if the password wasn’t hashed properly. With WebAuthn, if the server is compromised, the server only stores the public key, which requires the private key (stored in the user's device) to complete the authentication.
* It's easier for attackers to steal a user's password than to steal their device. Even if an attacker does steal a device, they still need to bypass the device's security (*ie* biometrics or facial recognition system).
* WebAuthn is easy to implement because it consists of [browser APIs that can be easily used by developers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API#examples).
* It offers a smooth authentication experience for users.

## How WebAuthn Works

> WebAuthn consists of three components: the authenticator (the user's device), the web browser, and the server (RP). In essence, the user initiates login via the browser, the browser then prompts the user to authenticate using their device (via biometrics or facial recognition, for example), and then if successful, the user is logged into the web application.

Reviewer: Facial recognition is a form of biometrics. I'd change this to be either:
* fingerprint or facial recognition
* biometrics or PIN

However, this is only a basic view of the process. To understand the process in more depth, you need to look at the two main flows of WebAuthn: the registration flow and the authentication flow. (They are called "ceremonies" in the specification.)

### The Registration Flow

> Before users can authenticate, their device(s) must be registered on the server of the RP. This way, the RP creates the credentials that the user can authenticate, and the process looks like this:

Reviewer: ```suggestion
Before users can authenticate, their device(s) must be registered on the server of the RP. This way, the RP creates the credentials that the user can use to authenticate, and the process looks like this:
```

It seems like there's a word or two missing from that line. Should it be something like "user can authenticate with" or "creates the authentication credentials for the user"?

{% include _image.liquid src="/assets/img/blogs/webauthn-terminology/registration-flow.png" alt="WebAuth Registration Flow diagram" class="img-fluid" figure=true %}

1. The user visits the web application of the RP and initiates registration.
2. The server creates a challenge key for registration, which is only used once.
3. The challenge key, along with the user's info, is sent to the user's device.

> 4. The user is prompted to authenticate themselves (this could be via biometrics, facial recognition system, or other forms depending on the device).

Reviewer: ```suggestion
4. The user is prompted to authenticate themselves (this could be via fingerprint, facial recognition, or other forms depending on the device).
```

> 5. After authentication success, the authenticator (the user's device) creates a new key pair (the public and private keys). 

Reviewer: I wouldn't use the term "authentication" here (or the previous step). The spec refers to this as an "authorization gesture."

6. The authenticator uses the private key to sign the challenge and then sends the public key and a credential identifier alongside the challenge over to the server.
7. The server validates the credibility of the private key by using the public key to evaluate the signed challenge.
8. The server associates the credentials with the user's account for future authentication purposes and redirects them to the application.

### The Authentication Flow

After a user's device is registered with the RP, the user can be authenticated. There are six steps involved in the WebAuthn authentication process:

> {% include _image.liquid src="/assets/img/blogs/webauthn-terminology/authentication-flow.png" alt="WebAuth Authentication Flow diagram" class="img-fluid" figure=true %}

Reviewer: I love the visuals, but communication with the authenticator is mediated by the browser. Readers may misinterpret the diagrams as meaning that the server communicates directly with the authenticator during these ceremonies.

Maybe rather than changing the visuals it's enough to make a note on the steps below like:
> The server (RP) creates and sends a unique challenge to the authenticator (via the browser).

Or perhaps it's clear to others that all of the interaction with the RP server goes through the browser because that's how you're browsing.

The following are the steps for this authentication process:

1. The user initiates login via the browser.
2. The server (RP) creates and sends a unique challenge to the authenticator.
3. The authenticator receives the challenge.

> 4. The authenticator prompts the user for device authorization (again, via biometrics, facial recognition system, etc.).

Reviewer: ```suggestion
4. The authenticator prompts the user for device authorization (again, via biometrics, PIN, etc.).
```

5. Upon successful authorization, the authenticator creates a cryptographic signature and sends it back to the server.

> 6. The server verifies the signature of the unique challenge that it sent to the user (in step 2). Upon successful verification, the user is logged into the web application.

Reviewer: ```suggestion
6. The server verifies the signature of the unique challenge that it sent to the user (in step 2) using the public key the server stored during registration. Upon successful verification, the user is logged into the web application.
```

## WebAuthn vs. FIDO 2

WebAuthn was developed by the [World Wide Web Consortium (W3C)](http://www.w3.org/), which is the international standards organization for the world wide web. Meanwhile, Fast Identity Online 2 (FIDO2) was created by [FIDO Alliance](https://fidoalliance.org/) with the aim of creating passwordless authentication. WebAuthn is a core component of [FIDO2 specifications](https://fidoalliance.org/fido2/).

> The FIDO Alliance decided to partner with W3C to make FIDO2 authentication a standard for the web. WebAuthn became a part of the FIDO2 specification, and a [set of browser APIs](https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API) was finalized, which is widely used today for passwordless authentication.

Reviewer: ```suggestion
The FIDO Alliance decided to partner with W3C to make FIDO2 authentication a standard for the web. WebAuthn became a part of the FIDO2 specification, and a [set of browser APIs](https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API) was finalized, which are widely used today for passwordless authentication.
```

I guess "is" could be referring to the "set" rather than "APIs," but this sounds more correct.

These APIs are still being developed and expanded, but the core WebAuthn APIs used for registration and authentication are probably set.

## Common WebAuthn Terminology

If you're looking to enter into the WebAuthn space, it can be helpful to understand the following terminology:

### Cross-Platform vs. Single-Platform

[Cross-platform authenticators](https://developers.yubico.com/WebAuthn/WebAuthn_Developer_Guide/Platform_vs_Cross-Platform.html.) (also known as roaming authenticators) are authenticators that can be used with another device, like a [YubiKey](https://www.yubico.com/), to perform authentication.

[Single-platform authenticators](https://developers.yubico.com/WebAuthn/WebAuthn_Developer_Guide/Platform_vs_Cross-Platform.html) (also known as platform authenticators) are authenticators that are built into the device, like Android pattern gestures or biometrics, iOS Face ID, or voice unlocks.

### User Presence vs. User Verification 

> User presence indicates the user's current state (*ie* whether someone is present). For instance, an authenticator may have a button a person pushes to indicate they are in possession of it. In this situation, a user is present, but as a relying party, you don't know which user. 

Reviewer: It may be worth mentioning that a test of user presence is required for all WebAuthn ceremonies.

> On the other hand, user verification is concerned with the authenticity of the user and verifies that the user is authenticated, for example by requiring a pin or facial recognition before they are allowed access.

Reviewer: ```suggestion
On the other hand, user verification is concerned with proving the identity of the user, for example by requiring a pin or facial recognition before they are allowed access.
```

### First Factor vs. Second Factor

First factor authentication means that the user is identified and authenticated. This is often done with a username and password. The username identifies the user and the password authenticates them. 

> A second factor authentication uses another factor for authentication. This second factor doesn't need to identify the user, since that was accomplished with the first factor. By requiring a second factor of authentication, the authenticating system is more certain of who is authenticating. A typical example is when a user tries to login to an application, and then gets an OTP sent to their email which they can use to complete authentication.

Reviewer: ```suggestion
A second factor authentication uses another factor for authentication. This second factor doesn't need to identify the user, since that was accomplished with the first factor. By requiring a second factor of authentication, the authenticating system is more certain of who is authenticating. A typical example is when a user tries to log in to an application, and then gets an OTP sent to their email which they can use to complete authentication.
```

### Biometric Recognition vs. Biometric Authenticator

Biometric recognition is automated identification of a user's physical or behavioral traits (*ie* fingerprints, voice unlocks, and facial recognition system). In contrast, biometric authenticators are the hardware that is capable of processing biometric recognitions (*ie* retina scans and fingerprint scans).

### Authorization Gesture

An authorization gesture is one or more physical movements by a user interacting with an authenticator as part of credential enrollment. For instance, this can include setting up Face ID scans on an iPhone where the user is required to rotate their head to capture different angles of their faces.

### Relying Party

> The relying party is the web application using WebAuthn to authenticate the user (*ie* your application, [FusionAuth](https://fusionauth.io/), or Auth0.

Reviewer: ```suggestion
The relying party is the web application using WebAuthn to authenticate the user (*ie* your application, [FusionAuth](https://fusionauth.io/), or Auth0).
```
